src/test/java/de/rototor/pdfbox/graphics2d/ChartFileBuilder.java

```diff
@@ -0,0 +1,187 @@
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.pdfbox.pdmodel.PDDocument;
+import org.apache.pdfbox.pdmodel.PDPage;
+import org.apache.pdfbox.pdmodel.PDPageContentStream;
+import org.apache.pdfbox.pdmodel.common.PDRectangle;
+import org.apache.pdfbox.pdmodel.font.PDType1Font;
+import org.apache.pdfbox.pdmodel.graphics.form.PDFormXObject;
+import org.apache.pdfbox.util.Matrix;
+import org.jfree.chart.ChartFactory;
+import org.jfree.chart.JFreeChart;
+import org.jfree.chart.axis.SymbolAxis;
+import org.jfree.chart.plot.PlotOrientation;
+import org.jfree.chart.plot.XYPlot;
+import org.jfree.chart.renderer.xy.GradientXYBarPainter;
+import org.jfree.chart.renderer.xy.XYBarRenderer;
+import org.jfree.data.xy.XYIntervalSeries;
+import org.jfree.data.xy.XYIntervalSeriesCollection;
+
+import de.rototor.pdfbox.graphics2d.PdfBoxGraphics2D;
+
+public class ChartFileBuilder
+{
+    private PDDocument  document = null;
+
+    private String filePath = null;
+
+    private PDPage              currentPage   = null;
+    private PDPageContentStream contentStream = null;
+
+    public ChartFileBuilder(String filePath)
+    {
+        this.filePath = filePath;
+
+        document = new PDDocument();
+
+        addPage();
+    }
+
+    public void addPage()
+    {
+        currentPage = new PDPage();
+        currentPage.setMediaBox(PDRectangle.A4);
+        document.addPage(currentPage);
+    }
+
+    public void go()
+    {
+        try
+        {
+            printStuff();
+
+            // TODO - If saveState is true ==> saveGraphicsState()
+            boolean saveState = true;
+            printChart(10, 10, 550, 550, buildDummyChart(1), saveState);
+            addPage();
+            printChart(10, 10, 550, 550, buildDummyChart(2), saveState);
+
+            document.save(new File(filePath));
+            document.close();
+            System.out.println(filePath + " - DONE");
+        }
+        catch(Exception e)
+        {
+            e.printStackTrace();
+        }
+    }
+
+    // --------------------------------------------------------------------
+
+    private void printStuff() throws IOException
+    {
+        openContentStream();
+
+        contentStream.setFont(PDType1Font.HELVETICA, 24);
+        contentStream.beginText();
+        contentStream.newLineAtOffset(220, 740);
+        contentStream.showText("TEST");
+        contentStream.endText();
+
+        contentStream.setLineWidth(0.7f);
+        contentStream.moveTo( 50, 700);
+        contentStream.lineTo(450, 700);
+        contentStream.stroke();
+
+        closeContentStream();
+    }
+
+    private JFreeChart buildDummyChart(int n)
+    {
+        // ------------------------------------------
+        // NB: big memory allocation!
+        Map<Integer, String> m = new HashMap<>();
+        for(int i=0; i<1_000_000; i++)
+        {
+            m.put(i, Long.toHexString(Double.doubleToLongBits(Math.random())));
+        }
+        m = null;
```

Comment:
As you loop 1_000_000 times here, the Java JIT optimizer will kick in and optimize this method. As the JIT optimizer can proof that m is never read again, it can and will optimize this assignment out. So you pin a massive amount of memory which can only be freed after buildDummyChart() has returned.

As locale variables can't be volatile (to force the assignment of null and avoid any other optimization on this variable) the only way to free memory here is to call m.clear(). As long as the method does not return, the map will be pinned, i.e. the GC can not free it. But it could free the object tree associated with the map if you call clear().

I've not tested this code, and i won't merge it. Try to fix your memory problems first. It's likely that you get in some corner cases in either JFreeChart or Apache PDFBox regarding the OOM handling. You may even have some funny interaction with the ZLib native code, which is used to write the compressed PDF stream.

- Set a breakpoint on any OOM Exception. Maybe JFreeChart or PDFBox tries to handle them and does something wrong here.
- As said, don't even think of assigning null to a locale variable. Won't help. If its a collection, clear() it.
- Consider swapping your calculation stuff to disk. E.g. you could use an embedded H2 database backed by a temporary db file. The size of your dataset would not matter in that case. H2 is very good for such stuff, and may be even faster for calcuation as you can create indexes on your data. Just throw away the H2 database and it's datafile after you finished the calculations. 
- Simply increase the Java process heap size...

You may expose some OOM handling problems in PDFBox or JFreeChart here. But the main problem is that you fill up your heap.